optimum/habana/transformers/models/mpt/modeling_mpt.py

```diff
@@ -312,13 +339,17 @@ def prepare_inputs_for_generation(
                 input_ids = input_ids[:, remove_prefix_length:]
             else:
                 input_ids = torch.index_select(input_ids, 1, token_idx - 1)
+
+                if bucket_internal and token_idx is not None:
+                    attention_mask = attention_mask[:, :cache_idx]
             # Converting back to tuples as it should be, so there's no type mismatch when calling graph
             past_key_values = tuple([tuple(kv) for kv in past_key_values])
         elif bucket_internal and token_idx is not None:
-            # KV cache is pre allocated with reuse cache or will be padded with bucket internal
-            # hence for the 1st token we can slice the inputs till token idx for the fwd pass.
+            # for the 1st token we can slice the inputs till token idx for the fwd pass.
             input_ids = input_ids[:, :token_idx]
             attention_mask = attention_mask[:, :token_idx]
+            # for the 1st token the cache_idx is None
+            cache_idx = token_idx
```

Comment:
why do you need to pass cache_idx for the first token gen? you can leave it as None

Reply:
When bucket internal is enabled, the input was sliced.  Therefore, the seq_length of key_states will be short than the length of kv_cache. If leave cache_idx as None, it will initialize past_key_value incorrectly according to [the code](https://github.com/Jing1Ling/optimum-habana/blob/86272355a3547504825c70a16c9439a4efab592a/optimum/habana/transformers/models/mpt/modeling_mpt.py#L93C12-L102C15).